### PR TITLE
Implement SSH RUN mount

### LINKED ISSUE #2835: SSH agent socket mount support while building images
/kind feature

**Description**
I have some dockerfiles where one of the steps requires to download a private git repository. I have been able to build the image using buildkit, by using some of their experimental features. Specifically, I have needed "--ssh" build argument and a small modification on the RUN step downloading the private repo, to make it have the following experimental syntax: "RUN --mount=type=ssh". From my limited knowledge, this is mainly exposing host's ssh agent socket to the build process so that it can be used on the needed RUN step. 

I have tried to replicate that procedure with podman, but I have not found any documentation whether podman supports these buildkit's experimental features/syntaxes or not. So, I would like to know if podman already supports it or if there is any plan to support it. In case it is already supported, I would like to know, if possible, where can I find the instructions to make it work.

I'm completely newbie with podman, so sorry for any inconvenience.

### PR DESCRIPTION
**Pull request by ashley-cui**
Allow ssh socket from host or certain ssh keys to be exposed to a
certain RUN instruction, but not any other instructions, as well as not
showing up in the final image.
This is done by spawining a new agent from buildah and mounting
the listening socket inside the run. SSH_AUTH_SOCK inside the container
will be set to the socket mountpoint. The defualt mountpoint is
/run/buildkit/ssh_agent.{i}

Signed-off-by: Ashley Cui <acui@redhat.com>

#### What type of PR is this?
/kind feature
#### What this PR does / why we need it:
Allow ssh socket from host or certain ssh keys to be exposed to a
certain RUN instruction, but not any other instructions, as well as not
showing up in the final image.
This is done by spawining a new agent from buildah and mounting
the listening socket inside the run. SSH_AUTH_SOCK inside the container
will be set to the socket mountpoint. The defualt mountpoint is
/run/buildkit/ssh_agent.{i}

#### How to verify it
buildah bud --ssh id=default . 

RUN --mount=type=ssh,id=default

#### Which issue(s) this PR fixes:
Fixes #2835

#### Does this PR introduce a user-facing change?

```release-note
Support SSH mounts in RUN instruction
```

### CODE REVIEW

**Comment by ashley-cui** [comment]
WIP because CI might get funky on this one since I'm creating SSH Keys. Also, if there's a better way than having to install openssh-client on every test run via the Dockerfile, let me know since that seems like something that can maybe be cached? I think there's also another SELinux bug here but I'll ask @rhatdan again when he comes back from the beach :)

**Comment by ashley-cui** [comment]
Hmm any ideas on why this test is flakey? 

```
[+1028s] # #|     FAIL: buildah bud --ssh default=/var/tmp/tmpaa73fbae7124310db96d4c35/my-dir1/sshkey --signature-policy /var/tmp/go/src/github.com/containers/buildah/tests/./policy.json -t sshimg -f /var/tmp/go/src/github.com/containers/buildah/tests/./bud/run-mounts/Dockerfile.ssh /var/tmp/go/src/github.com/containers/buildah/tests/./bud/run-mounts
[+1028s] # #| expected: 'SHA256:uv1/oDQWxtukcK+BLfIUfgQSdZiwHGWkBdA+6rNxZzE'
[+1028s] # #|   actual: 'STEP 1/3: FROM alpine'
[+1028s] # #|         > 'STEP 2/3: RUN apk add openssh-client'
[+1028s] # #|         > 'fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/main/x86_64/APKINDEX.tar.gz'
[+1028s] # #|         > 'fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/community/x86_64/APKINDEX.tar.gz'
[+1028s] # #|         > '(1/5) Installing openssh-keygen (8.1_p1-r0)'
[+1028s] # #|         > '(2/5) Installing ncurses-terminfo-base (6.1_p20190518-r2)'
[+1028s] # #|         > '(3/5) Installing ncurses-libs (6.1_p20190518-r2)'
[+1028s] # #|         > '(4/5) Installing libedit (20190324.3.1-r0)'
[+1028s] # #|         > '(5/5) Installing openssh-client (8.1_p1-r0)'
[+1028s] # #|         > 'Executing busybox-1.30.1-r2.trigger'
[+1028s] # #|         > 'OK: 10 MiB in 19 packages'
[+1028s] # #|         > 'STEP 3/3: RUN --mount=type=ssh,id=default ssh-add -l'
[+1028s] # #|         > '2048 SHA256:uv1/oDQWxtukcK+BLfIUfgQSdZiwHGWkBdA+6rNxZzE  (RSA)'
[+1028s] # #|         > 'COMMIT sshimg'
[+1028s] # #|         > 'Getting image source signatures'
[+1028s] # #|         > 'Copying blob sha256:03901b4a2ea88eeaad62dbe59b072b28b6efa00491962b8741081c5df50c65e0'
[+1028s] # #|         > 'Copying blob sha256:152eee7e8e237223cdc7ea6cd10d8f960e0d4daac597f68dece25c12e9d8a3d8'
[+1028s] # #|         > 'Copying config sha256:53059a04757a58aad0a03ed1f2fcc8b933789f2a6f0de02b3d7098795bd47f68'
[+1028s] # #|         > 'Writing manifest to image destination'
[+1028s] # #|         > 'Storing signatures'
[+1028s] # #|         > '--> 53059a04757'
[+1028s] # #|         > 'Successfully tagged localhost/sshimg:latest'
[+1028s] # #|         > '53059a04757a58aad0a03ed1f2fcc8b933789f2a6f0de02b3d7098795bd47f68'
```
the expected does show up as a substring of actual, but it seems flakey.

File: tests/bud.bats
Comment (ashley-cui) [inline tests/bud.bats:3326]: This line hangs in the test when I run it locally, if anyone has any idea why, let me know

File: pkg/sshagent/sshagent.go
Comment (rhatdan) [inline pkg/sshagent/sshagent.go:75]: Could you try adding

label.SetTaskLabel(processLabel)

**Comment by ashley-cui** [comment]
@containers/buildah-maintainers  PTAL :smiley:

File: pkg/sshagent/sshagent.go
Comment (rhatdan) [inline pkg/sshagent/sshagent.go:152]: buildkit?  Shouldn't this say buildah?
If you copied this code out of buildkit, you should give some credit to them in the header.

File: pkg/sshagent/sshagent.go
Comment (rhatdan) [inline pkg/sshagent/sshagent.go:213]: Don't begin errors with "Error" and also don't capitalize them.

File: pkg/sshagent/sshagent.go
Comment (rhatdan) [reply pkg/sshagent/sshagent.go:152]: Theoretically should these be errors.New()
Ditto for all of the other references to buildkit and errors.Errorf.

File: run_linux.go
Comment (rhatdan) [inline run_linux.go:416]: Since there looks like the amount of options might expand and the return object might increase, should we create a return struct?

**Review by rhatdan — APPROVED** [review]
Looking good, just a few problems.

File: docs/buildah-bud.md
Comment (TomSweeneyRedHat) [inline docs/buildah-bud.md:508]: No pronouns, please!
```suggestion
The socket path can be left empty to use the value of `default=$SSH_AUTH_SOCK`
```

File: pkg/cli/common.go
Comment (TomSweeneyRedHat) [inline pkg/cli/common.go:226]:
```suggestion
	fs.StringArrayVar(&flags.SSH, "ssh", []string{}, "SSH agent socket or keys to expose to the build. (format: default|<id>[=<socket>|<key>[,<key>]])")
```

File: pkg/sshagent/sshagent.go
Comment (TomSweeneyRedHat) [inline pkg/sshagent/sshagent.go:38]: nit
```suggestion
// newAgentServerKeyring creates a new agent from scratch and adds keys
```

File: run.go
Comment (TomSweeneyRedHat) [inline run.go:143]:
```suggestion
	// SSH is the available ssh agents to use in a RUN
```

File: run_linux.go
Comment (TomSweeneyRedHat) [reply run_linux.go:416]: YES!  I always get leary when a function takes or returns more than 3 or 4 params, too easy to mess up when calling.

File: run_linux.go
Comment (TomSweeneyRedHat) [inline run_linux.go:2335]:
```suggestion
				// Count is needed as the default destination of the ssh socket inside the container is  /run/buildkit/ssh_agent.{i}
```

Comment (TomSweeneyRedHat) [reply run_linux.go:2335]: Oops, nope, too quick with my red pen on this one.  ;^)

File: run_linux.go
Comment (TomSweeneyRedHat) [inline run_linux.go:2443]:
```suggestion
// getSSHMount parses the --mount type=ssh flag in the Containerfile, checks if there's an ssh source provided, and creates and starts an ssh-agent to be forwarded into the container
```

File: run_linux.go
Comment (TomSweeneyRedHat) [inline run_linux.go:2453]: If you don't have an equal sign in the token, you'll get some unhappiness.  Might want to do a len check on kv to be completely safe.

File: pkg/sshagent/sshagent.go
Comment (ashley-cui) [reply pkg/sshagent/sshagent.go:152]: Does something like 
// readOnlyAgent and its functions originally from moby/buildkit 
work?

File: run_linux.go
Comment (ashley-cui) [reply run_linux.go:416]: would it be a better idea to return a function that cleanups all these mounts? since the only reason we're returning them is to defer the cleanup later? or do you think a struct is better for this case

**Comment by ashley-cui** [comment]
Here's the strange error that I'm getting: 
```
STEP 4/4: RUN --mount=type=ssh,id=default ssh-add -l; ssh git@github.com
3072 SHA256:UP+1IIyKqBfXHqlLaE9JO8L088FKOy+M3aHyf+gBgK4 acui@fedora (RSA)
256 SHA256:YkQgBsosvLxPjxjLVl18XIEMytqodMIYYFDQg9aVQcY root@fedora (ED25519)
256 SHA256:svQtLIoRuxW7kdSWlwQwWCuybH/DchNZPopAGBpzEMM acui@redhat.com (ED25519)
Pseudo-terminal will not be allocated because stdin is not a terminal.
ERRO[0005] error serving agent: read unix /var/tmp/.buildah-ssh-sock421179333/ssh_auth_sock->@: use of closed network connection 
sign_and_send_pubkey: signing failed for RSA "acui@fedora" from agent: communication with agent failed
```

It seems strange because ssh-add doesn't fail but the actual ssh command does. both should be using  the sock but it only throws failed on the ssh and not the ssh-add

File: run_linux.go
Comment (TomSweeneyRedHat) [reply run_linux.go:416]: If the defer works for you, that's fine.  My deal is just in general I don't like to see 3 or 4 params being passed back or forth.  YMMV.

**Comment by rhatdan** [comment]
I think the problem could be about Pseudo-terminal will not be allocated because stdin is not a terminal. could you try with a scp rather then an ssh?

**Comment by ashley-cui** [comment]
I found it! It was just that my timeout was too short. I changed to to 500ms.

**Comment by ashley-cui** [comment]
Updated and fixed @containers/buildah-maintainers PTAL :)

**Review by rhatdan — APPROVED** [review]
LGTM

**Comment by TomSweeneyRedHat** [comment]
LGTM
nice work @ashley-cui , take tomorrow off!

**Comment by ashley-cui** [comment]
I'll let @rhatdan decide that one, but we have to be careful when this gets into podman, since this won't work with remote build

**Comment by rhatdan** [comment]
No, let's sit on this for a while.

**Pull request merged**
